### PR TITLE
Report Detailed Error Reports to disk

### PR DESCRIPTION
**Pull request by stefanpenner**
This aims to restore debuggability we lost from https://github.com/ember-cli/console-ui/pull/14 (but maintain the improved day 2 day ergonomics improved by #14)

the error report ends up having a filepath of the following format:
```
<tmpdir>/error.dump.<checksum-of-the-repo>.log
```

Why?

* cleanup after themselves
* doesn't conflict, and doesn't pollute cwd
* configurable via `TMPDIR`

Alternatives:

* local `tmp` or `log`
* system `log` ?

Example filepath:

```
/private/var/folders/4r/whc65vwj1xggvvky3yy1cp9m000mw4/T/error.dump.1c7e928964050f90db424cbdaef761ec.log
```

the error reports end up looking like:

```
=================================================================================

ENV Summary:

  TIME: Fri Nov 24 2017 18:54:14 GMT-0800 (PST)
  TITLE: node
  ARGV:
  - /Users/spenner/src/stefanpenner/dotfiles/.config/node/versions/node-v8.9.0-darwin-x64/bin/node
  EXEC_PATH: /Users/spenner/src/stefanpenner/dotfiles/.config/node/versions/node-v8.9.0-darwin-x64/bin/node
  PATH:
  - /Users/spenner/.fzf/bin
  - /Users/spenner/.cargo/bin
  - /Users/spenner/.config/node/default/bin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  - /usr/local/linkedin/bin
  - /export/content/linkedin/bin
  PLATFORM: darwin x64
  VERSIONS:
  - ares: 1.10.1-DEV
  - cldr: 31.0.1
  - http_parser: 2.7.0
  - icu: 59.1
  - modules: 57
  - nghttp2: 1.25.0
  - node: 8.9.0
  - openssl: 1.0.2l
  - tz: 2017b
  - unicode: 9.0
  - uv: 1.15.0
  - v8: 6.1.534.46
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: [undefined]
  - codeFrame: [undefined]
  - errorMessage: 
  - errorType: [undefined]
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
  - message: 
  - name: Error
  - nodeAnnotation: [undefined]
  - nodeName: [undefined]
  - originalErrorMessage: [undefined]
  - stack: Error
    at repl:1:14
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:441:10)
    at emitOne (events.js:121:20)
    at REPLServer.emit (events.js:211:7)
    at REPLServer.Interface._onLine (readline.js:282:10)
    at REPLServer.Interface._line (readline.js:631:8)

=================================================================================

```

Pictures:

screen shot 2017-11-24 at 7 06 58 pm

### CODE REVIEW

**Comment by stefanpenner** [comment]
cc @josemarluedke

File: tests/unit/write-error-test.js
Comment (rwjblue) [inline tests/unit/write-error-test.js]: Odd spacing here (needs a space after =)

**Review by rwjblue — APPROVED** [review]

**Comment by stefanpenner** [comment]
I think this should also accept additional meta data from the consumer. Allowing things like ember-cli to add additional info (like its version etc.)

Maybe something like:
```js

new UI({
  onErrorReport(report) { return updatedReported; }
})
```

**Comment by rwjblue** [comment]
Ya, sounds great. Info on ember version, other addons etc would be awesome

File: lib/summarize-process.js
Comment (twokul) [inline lib/summarize-process.js]: `summarizeProcess` w/o the "p"?

File: lib/index.js
Comment (twokul) [inline lib/index.js]: not 100% sure but will "/" work on Windows? or maybe we should use something like `path.join(tmpdir, 'error.dump.${md5}.log')`?

**Review by twokul — APPROVED** [review]

**Comment by twokul** [comment]
@stefanpenner thanks for catching this, Stef!

File: lib/index.js
Comment (stefanpenner) [reply lib/index.js]: all `fs.*` methods automagically transform `/` to the appropriate slash. Our code can be all written posix style.

**Comment by stefanpenner** [comment]
Updated output based on @rwjblue  suggestion:

```
=================================================================================

ENV Summary:

  TIME: Fri Nov 24 2017 21:46:19 GMT-0800 (PST)
  TITLE: node
  ARGV:
  - /Users/spenner/src/stefanpenner/dotfiles/.config/node/versions/node-v8.9.0-darwin-x64/bin/node
  EXEC_PATH: /Users/spenner/src/stefanpenner/dotfiles/.config/node/versions/node-v8.9.0-darwin-x64/bin/node
  TMPDIR: /var/folders/4r/whc65vwj1xggvvky3yy1cp9m000mw4/T
  SHELL: /usr/local/bin/fish
  PATH:
  - /Users/spenner/.cargo/bin
  - /Users/spenner/.config/node/default/bin
  - /Users/spenner/.fzf/bin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  - /usr/local/linkedin/bin
  - /export/content/linkedin/bin
  PLATFORM: darwin x64
  FREEMEM: 4938248192
  TOTALMEM: 17179869184
  UPTIME: 992560
  LOADAVG: 1.45361328125,1.4130859375,1.4765625
  CPUS:
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  ENDIANNESS: LE
  VERSIONS:
  - ares: 1.10.1-DEV
  - cldr: 31.0.1
  - http_parser: 2.7.0
  - icu: 59.1
  - modules: 57
  - nghttp2: 1.25.0
  - node: 8.9.0
  - openssl: 1.0.2l
  - tz: 2017b
  - unicode: 9.0
  - uv: 1.15.0
  - v8: 6.1.534.46
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: [undefined]
  - codeFrame: [undefined]
  - errorMessage: omg the message
  - errorType: [undefined]
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
  - message: omg the message
  - name: Error
  - nodeAnnotation: [undefined]
  - nodeName: [undefined]
  - originalErrorMessage: [undefined]
  - stack: Error: omg the message
    at [eval]:1:33
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at Object.runInThisContext (vm.js:139:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:635:30)
    at evalScript (bootstrap_node.js:462:27)
    at startup (bootstrap_node.js:163:9)
    at bootstrap_node.js:608:3
```

**Comment by stefanpenner** [comment]
@rwjblue  / @twokul  / @josemarluedke  if you guys can give ^ a whirl (do some NPM linking to make sure the ergonomics are good) in the next day or so that would be sweet. I'm traveling tomorrow, so my time online will be sporadic.

**Comment by stefanpenner** [comment]
@rwjblue finally got around to fixing the CI issues. I haven't had time to do a full manual integration test with ember-cli (and make sure a-few common errors feel good). I think we should hold off until someone does this. I can likely do this in the AM... but I welcome others, as they can provide additional feedback.

**Comment by rwjblue** [comment]
Sounds good. Definitely want to make sure things feel nice. 

Unsure what time I’ll have tonight / tomorrow, but if someone hasn’t gotten to it by Mon I’ll jump on it...

**Comment by josemarluedke** [comment]
@stefanpenner This is awesome. I submitted #22 to fix writing the the file to disk. Thank you for working on this.

**Comment by stefanpenner** [comment]
> Unsure what time I’ll have tonight / tomorrow, but if someone hasn’t gotten to it by Mon I’ll jump on it...

It's ready for you!

also, I added a high level `ui.writeError` test

File: tests/unit/index-test.js
Comment (stefanpenner) [inline tests/unit/index-test.js:87]: @rwjblue extra tests

**Comment by twokul** [comment]
@stefanpenner I'm in Europe atm, back on Thursday. I can try it out then but changes looks good, no objections to 🐑

@rwjblue @josemarluedke thoughts?

**Review by hjdivad — APPROVED** [review]
LGTM (pending end-to-end test)

**Comment by stefanpenner** [comment]
@rwjblue  / @hjdivad et al, thoughts on including `package.json` (i suspect lock files might be a tad to big, but they may also be good...)

**Comment by stefanpenner** [comment]
- [ ] needs npm version or yarn version

**Comment by hjdivad** [comment]
> @rwjblue / @hjdivad et al, thoughts on including package.json (i suspect lock files might be a tad to big, but they may also be good...)

I think including `package.json` is good.  I'd be 👍  on including lockfiles by default as well (maybe with a size cutoff / option) but even this + `package.json` goes a long way.

**Comment by stefanpenner** [comment]
manually tested, looks good. Always more improvements, but we can likely land.

**Comment by stefanpenner** [comment]
example: 

screen shot 2017-12-21 at 9 09 19 pm

**Comment by stefanpenner** [comment]
released as v2.1.0 🎉

**Pull request merged**
